### PR TITLE
Define Mapping of Arbitrary Data to OTLP AnyValue

### PR DESCRIPTION
**Pull request by tigrannajaryan**
The mapping is necessary to correctly implement Logging Library SDKs such that
values mapped are unambiguous, consistent across languages and implementations.

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@jkwatson @zenmoto PTAL.

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (yurishkuro) [inline text/logs/0178-mapping-to-otlp-anyvalue.md:113]: "Ensuring uniqueness" doesn't specify the method of doing so. What if instead we simply refer to the following text on handling non-unique keys?

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (yurishkuro) [inline text/logs/0178-mapping-to-otlp-anyvalue.md:118]: Add header "duplicate keys"

**Review by yurishkuro — APPROVED** [review]

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (jkwatson) [inline text/logs/0178-mapping-to-otlp-anyvalue.md:4]: Do we want to call out that this is specifically for the log datamodel's "body"?

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (jkwatson) [inline text/logs/0178-mapping-to-otlp-anyvalue.md:42]: Hmm. the specific example of `enum` in Java is problematic, as we have been implored repeatedly not to assume that the ordinal values will be stable or useful for serialization purposes. I think Java's enums should be mapped to their name, perhaps prefixed with the actual enum Type in some way.

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (jkwatson) [inline text/logs/0178-mapping-to-otlp-anyvalue.md:102]: If the array is heterogenous (like `Object[]` in Java), is this still true? Are we ok with the contents of the AnyValue array type being hetergenous?

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/logs/0178-mapping-to-otlp-anyvalue.md]: Changed to prefer symbolic name over ordinal number.

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (jsuereth) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:100]: +1

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/logs/0178-mapping-to-otlp-anyvalue.md]: Reworded and added examples.

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/logs/0178-mapping-to-otlp-anyvalue.md]: Made "Associative Arrays With Non-Unique Keys" a separate section.

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-approvers Please review. This is needed by Logging SDK for now but there is no reason this cannot be extended to Span, Resource or SpanEvent attributes in the future if we decide so. It is best that we make it future-proof and ensure it is reasonable for all signals (even if we don't need it now).

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-logs-approvers please review.

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (pmm-sumo) [inline text/logs/0178-mapping-to-otlp-anyvalue.md:81]: Should we provide some guidelines on how to interpret the bytes sequence? Maybe we could add a semantic convention for an attribute that could store information about the data format (e.g. following [media type](https://en.wikipedia.org/wiki/Media_type) or using [Body Meta-Data](https://docs.google.com/document/d/1ZExye1lW43owwaxcbjOvl0P2qER-UaYd_MyxItc2h0k/edit#heading=h.lsix8inep3nb))

We could introduce a standard attribute name for body (e.g. `body.media.type`) and a convention for attributes (e.g. for attribute `foo`, bytes could be stored into `foo.bytes` while type into `foo.media.type`). That might add some complexity but also provide the much needed information on how the data should be interpreted

Comment (kumoroku) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:81]: unsurprisingly, I like the idea of having a way to annotate what is supposed to be found in a byte array (or even, depending on implementation in a string) :)

I think that we could turn this into another work item as suitable (another OTEP?), independently from approving this document.

**Review by kumoroku — APPROVED** [review]

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:81]: I agree. I think we can have a semantic convention for recording the Body's media type.

Comment (pmm-sumo) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:81]: Should we make it part of this oteps, a separate one or just add to semantic conventions eventually?

Comment (tigrannajaryan) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:81]: I think we can simply submit a separate semantic convention PR against spec repo. Not sure if an OTEP is necessary, but if we think the description of the semantic convention is going to be non-trivial maybe do an OTEP.

**Review by pmm-sumo — APPROVED** [review]

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (djaglowski) [inline text/logs/0178-mapping-to-otlp-anyvalue.md:181]: Should we define an expected mapping for a set? 

[Many languages](https://en.wikipedia.org/wiki/Set_(abstract_data_type)#Language_support) have a representation, so it seems we can either establish an imperfect but consistent mapping, or expect to have a variety of inconsistent mappings.

Perhaps the best representation would be as a `kvlist_value`, where each `key` represents an item in the set and each `value` is `empty`.

Comment (pmm-sumo) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:181]: > Should we define an expected mapping for a set?
> 
> [Many languages](https://en.wikipedia.org/wiki/Set_(abstract_data_type)#Language_support) have a representation, so it seems we can either establish an imperfect but consistent mapping, or expect to have a variety of inconsistent mappings.

That's a good point

> Perhaps the best representation would be as a `kvlist_value`, where each `key` represents an item in the set and each `value` is `empty`.

This would limit the `key` type to `string` though, wouldn't it? Perhaps we could simply recommend using Arrays for sets?

Comment (djaglowski) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:181]: > This would limit the type though as key must be string

Yeah, that's true. 

> recommend using Arrays for sets

I agree with this.

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:70]: I am unable to find which version of Unicode Protobuf adheres to, so this probably needs to remain unspecified for now.

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:52]: >should have a unique representation within the resulting set of bytes_value's.

If it is not possible to obtain the symbolic name or the ordinal value of the enum then we are probably in serious trouble. In such circumstances it is not clear if it is possible at all to construct a unique representation. I am not sure requiring it is going to be helpful. Worst case we should be ready to see something like `unknown` in OTLP, regardless of the input enum's value.

File: text/logs/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/logs/0178-mapping-to-otlp-anyvalue.md:191]: Added a section for Sets.

**Comment by tigrannajaryan** [comment]
Lint failures will be fixed by https://github.com/open-telemetry/oteps/pull/196 which needs to be merged before this PR.

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-approvers please review.

**Review by djaglowski — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-trace-approvers please review.

**Review by jsuereth — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-approvers last chance to review/comment/object before I merge this and start moving to the spec. This is going to apply to all signals.

**Review by dashpole — APPROVED** [review]

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (MrAlias) [inline text/0178-mapping-to-otlp-anyvalue.md:75]:
```suggestion
String values which are not valid Unicode sequences SHOULD be converted to
```

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (MrAlias) [inline text/0178-mapping-to-otlp-anyvalue.md:70]: Does an encoding of Unicode need to be specified?

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (MrAlias) [inline text/0178-mapping-to-otlp-anyvalue.md:18]: With the goal being to have consistency across implementation do we need to explain when it is valid to not follow this recommendation? Or, should this be made into a requirement?

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (MrAlias) [inline text/0178-mapping-to-otlp-anyvalue.md:32]: This mapping of many types into a single "string_value" type means this transformation is not going to be reversible.

Based on the PR description, when we say we want an unambiguous transformation we are only talking about one direction, right?

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/0178-mapping-to-otlp-anyvalue.md:70]: I don't think so. OTLP Protobuf requires UTF8, but the source from which we map can be any other Unicode encoding and can be converted to UTF8. So I think saying "valid Unicode" is a sufficient restruction.

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/0178-mapping-to-otlp-anyvalue.md:18]: Hard to tell what such explanation can look like. I used the "SHOULD" qualifier because I can't prove these recommendations are applicable to every possible use case.

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/0178-mapping-to-otlp-anyvalue.md:32]: Correct. Not reversible. Only "unambiguous" in the sense that it attempts to clearly say what to do. Maybe it is not the right word to use.

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (tigrannajaryan) [reply text/0178-mapping-to-otlp-anyvalue.md]: Fixed.

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (MrAlias) [reply text/0178-mapping-to-otlp-anyvalue.md:18]: It seems like there is a compatibility issue with implementations not conforming to this specification, different implementations for the same operations would produce different telemetry. This seems to warrant requiring these statements instead of recommending them.

This isn't a blocker. I just want to make sure we understand the implications. It is also important to keep in mind that softening a requirement to a recommendation is backwards compatible. The other way is not.

**Review by MrAlias — APPROVED** [review]
:rocket:

File: text/0178-mapping-to-otlp-anyvalue.md
Comment (zenmoto) [inline text/0178-mapping-to-otlp-anyvalue.md:88]: It might be prudent to explicitly call out at this level of detail that composite values can contain values of differing types, though it also does seem clear from the type definitions.

**Pull request merged**
